File: bench/kernels/alias_daxpy/ref.c
/* alias_daxpy: y[i] += a*x[i] over 4096 doubles, `ops` sweeps.
 * The aliasing kernel: C only vectorizes this freely with hand-written
 * restrict — exactly the fact wolf's type system will prove for free (D3).
 * Protocol: argv[1]=ops; prints {"ns":<total>,"ops":<ops>} on stdout. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 4096

static void daxpy(double *restrict y, const double *restrict x, double a) {
    for (int i = 0; i < N; i++) y[i] += a * x[i];
}

volatile double sink;

int main(int argc, char **argv) {
    uint64_t ops = argc > 1 ? strtoull(argv[1], 0, 10) : 1000;
    static double x[N], y[N];
    for (int i = 0; i < N; i++) { x[i] = (double)i * 0.5; y[i] = (double)i; }
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (uint64_t k = 0; k < ops; k++) daxpy(y, x, 1.000001);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    sink = y[N - 1];
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%llu}\n",
           (unsigned long long)ns, (unsigned long long)ops);
    return 0;
}

File: bench/kernels/alias_daxpy/ref.rs
//! alias_daxpy = a1-triad (family A), Rust reference: `&mut [f64]` is
//! noalias by construction, so Rust gets for free what C needs `restrict`
//! for — the same fact wolf's param modes carry.
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;
use std::time::Instant;

const N: usize = 4096;

fn daxpy(y: &mut [f64; N], x: &[f64; N], a: f64) {
    for i in 0..N {
        y[i] += a * x[i];
    }
}

fn main() {
    let ops: u64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(20000);
    let mut y = [0.0f64; N];
    let mut x = [0.0f64; N];
    let mut v = 0.0f64;
    let mut w = 0.0f64;
    for i in 0..N {
        x[i] = v;
        y[i] = w;
        v += 0.5;
        w += 1.0;
    }
    let t = Instant::now();
    for _ in 0..ops {
        daxpy(&mut y, &x, 1.000001);
    }
    let ns = t.elapsed().as_nanos();
    let sink = black_box(y[N - 1]);
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink:?}}}",
        ops * N as u64
    );
}
